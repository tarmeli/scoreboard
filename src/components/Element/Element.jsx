import React from "react";

const Element = ({
    className,
    tagName,
    tag: Tag = tagName,
    children,
    ...props
}) => (
    <Tag className={className} {...props}>
        {children}
    </Tag>
);

export default Element;
