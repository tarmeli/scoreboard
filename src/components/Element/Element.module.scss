.fitContent {
  width: fit-content;
}

.halfWidth {
  width: 50%;
}

.fullWidth {
  width: 100%;
}

.flexColumn {
  flex-direction: column;
}

.flexContainer {
  display: flex;
}

.centerContent {
  align-self: center;
  width: 33%
}
