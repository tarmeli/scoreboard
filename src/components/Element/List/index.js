import Li from "./Li/Li";
import Ol from "./Ol/Ol";
import Ul from "./Ul/Ul";

export { Li, Ol, Ul };
