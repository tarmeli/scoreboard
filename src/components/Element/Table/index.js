import Tbody from "./Tbody/Tbody";
import Td from "./Td/Td";
import Th from "./Th/Th";
import Thead from "./Thead/Thead";
import Tr from "./Tr/Tr";

export { Tbody, Td, Th, Thead, Tr };
