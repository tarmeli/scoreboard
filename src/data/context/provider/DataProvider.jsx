import React, { createContext, useContext, useEffect, useState } from "react";
import callApi from "../../utils/callApi/callApi";
import { dataModel } from "../../model/dataModel";

const DataContext = createContext({
  data: dataModel,
  isLoading: false,
  error: { isError: false, message: "" },
  callApiFor: () => {},
});

export const dependencies = {
  callApi,
};

const DataProvider = ({ route, children }) => {
  const [data, setData] = useState(dataModel);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState({
    isError: false,
    message: "",
  });

  const callApiFor = async (route, method) => {
    const { callWasSuccessful, data, error } = await dependencies.callApi(
      route,
      method
    );

    if (callWasSuccessful) {
      setData(data);
    } else {
      setError({
        isError: true,
        message: `${route} API call was not successful, because of ${error}`,
      });
    }
  };

  useEffect(() => {
    setIsLoading(true);
    callApiFor(route).then(() => setIsLoading(false));
  }, []);

  return (
    <DataContext.Provider value={{ data, isLoading, error, callApiFor }}>
      {children}
    </DataContext.Provider>
  );
};

export const useData = () => useContext(DataContext);

export default DataProvider;
