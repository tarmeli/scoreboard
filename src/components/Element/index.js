import A from "./A/A";
import Button from "./Button/Button";
import Div from "./Div/Div";
import Nav from "./Nav/Nav";
import Input from "./Input/Input";
import Table from "./Table/Table";

export { A, Button, Div, Nav, Input, Table };
