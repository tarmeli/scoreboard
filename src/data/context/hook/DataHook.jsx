import React from "react";
import { useData } from "../provider/DataProvider";
import LoadingTentative from "../../../components/LoadingTentative/LoadingTentative";
import ErrorTentative from "../../../components/ErrorTentative/ErrorTentative";

const DataHook = ({ children: toChild }) => {
  const {
    data,
    isLoading,
    error: { isError, message },
    callApiFor,
  } = useData();

  return (
    <LoadingTentative isLoading={isLoading}>
      <ErrorTentative isError={isError} message={message}>
        {toChild({ data, callApiFor })}
      </ErrorTentative>
    </LoadingTentative>
  );
};

export default DataHook;
