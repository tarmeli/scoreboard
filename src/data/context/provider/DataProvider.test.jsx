import React, { useState as useStateMock } from "react";
import { mount } from "enzyme";
import asyncFn from "@async-fn/jest";

import DataProvider, { dependencies } from "./DataProvider";
import DataHook from "../hook/DataHook";
import { Div } from "../../../components/Element/";

const Consumer = ({ data: { content } }) => <Div>{content}</Div>;

describe("DataProvider", () => {
  let component;
  let setStateMock;
  let callApiMock;

  beforeEach(() => {
    setStateMock = jest.fn();
    callApiMock = asyncFn();

    dependencies.callApi = callApiMock;
    useStateMock.mockImplementation((init) => [init, setStateMock]);

    component = mount(
      <DataProvider route="some-route">
        <DataHook>{({ data }) => <Consumer data={data} />}</DataHook>
      </DataProvider>
    );
  });

  it("when api call resolves with some content, renders", async () => {
    await callApiMock.resolve({
      callWasSuccessful: true,
      data: { content: "Some content" },
      error: { isError: false, message: "" },
    });

    expect(component).toMatchSnapshot();
  });

  it("when api call returns data, sets correct data", async () => {
    await callApiMock.resolve({
      callWasSuccessful: true,
      error: { isError: false, message: "" },
      data: { content: "Some content" },
    });

    expect(setStateMock).toHaveBeenCalledWith({
      content: "Some content",
    });
  });

  it("when api call fails, sets correct error", async () => {
    await callApiMock.resolve({
      callWasSuccessful: false,
      error: { isError: true, message: "some error" },
    });

    expect(setStateMock).toHaveBeenCalledWith({
      isError: true,
      message: "some-route API call was not successful, because of some error",
    });
  });
});
