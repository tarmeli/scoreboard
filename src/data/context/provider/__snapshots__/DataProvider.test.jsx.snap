// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataProvider when api call resolves with some content, renders 1`] = `
<DataProvider
  route="some-route"
>
  <DataHook>
    <LoadingTentative
      isLoading={false}
    >
      <Div>
        <Element
          tagName="div"
        >
          <div>
            <ErrorTentative
              isError={false}
              message=""
            >
              <Div>
                <Element
                  tagName="div"
                >
                  <div>
                    <Consumer
                      data={
                        Object {
                          "content": "",
                          "players": Array [],
                        }
                      }
                    >
                      <Div>
                        <Element
                          tagName="div"
                        >
                          <div />
                        </Element>
                      </Div>
                    </Consumer>
                  </div>
                </Element>
              </Div>
            </ErrorTentative>
          </div>
        </Element>
      </Div>
    </LoadingTentative>
  </DataHook>
</DataProvider>
`;
